068-performance-testing.md

```diff
@@ -0,0 +1,319 @@
+        // Create topics
+        KafkaTopicScalabilityUtils.createTopicsViaK8s(testStorage.getNamespaceName(), testStorage.getClusterName(), topicNamePrefix,
+                numberOfTopics, 12, 3, 1);
+
+        KafkaTopicScalabilityUtils.waitForTopicsReady(testStorage.getNamespaceName(), topicNamePrefix, numberOfTopics);
+
+        endTime = System.nanoTime();
+        createTopicsTimeNs = endTimeNs - startTimeNs;
+        LOGGER.info("Time taken to create {} topics: {} ns", numberOfTopics, createTopicsTimeNs);
+
+        endTimeWholeNs = System.nanoTime();
+        totalTimeWholeNs = endTimeWholeNs - startTimeNs;
+
+        LOGGER.info("Time taken to create {} topics and send and recv: {} messages: {} ns", numberOfTopics, NUMBER_OF_MESSAGES, totalTimeWholeNs);
+        
+        // Start measuring time for deletion of all topics
+        long deletionStartTimeNs = System.nanoTime();
+
+        // 2b. Exercise: Deletion of KafkaTopics
+        ResourceManager.getInstance().deleteResourcesOfType(KafkaTopic.RESOURCE_KIND);
+
+        long deletionEndTimeNs = System.nanoTime();
+        totalDeletionTimeNs = deletionEndTimeNs - deletionStartTimeNs;
+        LOGGER.info("Time taken to delete {} topics: {} ns", numberOfTopics, totalDeletionTimeNs);
+
+        endTimeWholeNs = System.nanoTime();
+        totalTimeWholeNs = endTimeWholeNs - startTimeNs;
+
+        LOGGER.info("Total time taken to create {} topics, send and receive {} messages, and delete topics: {} ns", numberOfTopics, NUMBER_OF_MESSAGES, totalTimeWholeNs);
+    } finally {
+        // 4a. teardown: Gracefully stop the topic-operator thread
+        // 4b: teardown: Now, it's safe to log performance data as the collection thread has been stopped
+        PerformanceReporter.logPerformanceData(
+                numberOfTopics, numberOfClientInstances, NUMBER_OF_MESSAGES, createTopicsTimeNs, totalSendAndRecvTimeNs, 
+                totalDeletionTimeNs, totalTimeWholeNs, this.testStorage, ACTUAL_TIME, Environment.PERFORMANCE_DIR, 
+                maxBatchSize, maxBatchLingerMs, this.topicOperatorPollingThread.getMetricsHistory()
+        );
+    }
+}
+```
+
+Note that this is just an one possible example. There is no need to use @ParametrizedTest.
+We could also use simple @Test to edge performance cases.
+
+### Gathered metrics
+
+During test execution we need to scrape metrics from the `SUT` pods. 
+Therefore, one way how to tackle such problem is that we could use a `Thread`, which would run a infinite loop, where we would scrape all related metrics f.e.:
+```java
+@Override
+public void run() {
+    LOGGER.info("Thread started with selector: {}", this.selector);
+
+    while (!Thread.currentThread().isInterrupted()) {
+        this.topicOperatorMetricsCollector.collectMetricsFromPods();
+        // record specific time when metrics were collected
+        Long timeWhenMetricsWereCollected = System.nanoTime();
+
+        Map<String, List<Double>> currentMetrics = new HashMap<>();
```

Comment:
Shouldn't do something with `currentMetrics` once you're populated it?

Reply:
We do...this is just for brevity and illustration that we would create a separate Thread within test execution where we would gather metrics. 
For instance:
1. We calculate derived metrics in the loop (I know this could be calculated within that loop)
2. (main one) We create a file for each metric so we loop through them in the `Reporter` class.

```java
metricsHistory.forEach((time, metrics) -> {
            // Including milliseconds in the time format
            DateTimeFormatter dateTimeFormatter = DateTimeFormatter.ofPattern("yyyy-MM-dd-HH-mm-ss-SSS");
            dateTimeFormatter = dateTimeFormatter.withZone(ZoneId.systemDefault());
            String formattedTime = dateTimeFormatter.format(Instant.ofEpochMilli(time));

            metrics.forEach((metricName, values) -> {
                String content = String.format("Timestamp: %s, Values: %s%s", formattedTime, values, newLine);

                // Using just the metric name for the file name
                String fileName = metricName + ".txt";
                Path metricFilePath = alicePathDir.resolve(fileName);

                // Appending to the file if it exists, or creating a new one if it doesn't
                try (BufferedWriter writer = Files.newBufferedWriter(metricFilePath,
                        StandardCharsets.UTF_8,
                        Files.exists(metricFilePath) ? StandardOpenOption.APPEND : StandardOpenOption.CREATE)) {
                    writer.write(content);
                } catch (IOException e) {
                    LOGGER.error("Failed to write metrics to file", e);
                }
            });
        });
```